File: core/src/main/scala/pl/ostrzyciel/jelly/core/ProtoEncoder.scala
package pl.ostrzyciel.jelly.core

import pl.ostrzyciel.jelly.core.proto.*

import java.math.BigInteger
import scala.collection.mutable.ListBuffer

/**
 * Stateful encoder of a protobuf RDF stream.
 *
 * This class supports all stream types and options, but usually does not check if the user is conforming to them.
 * It will, for example, allow the user to send generalized triples in a stream that should not have them.
 * Take care to ensure the correctness of the transmitted data, or use the specialized wrappers from the stream package.
 * @param options options for this stream
 */
abstract class ProtoEncoder[TNode >: Null <: AnyRef, TTriple, TQuad, TQuoted](val options: RdfStreamOptions):
  // *** 1. THE PUBLIC INTERFACE ***
  // *******************************
  /**
   * Add an RDF triple statement to the stream.
   * @param triple triple to add
   * @return iterable of stream rows
   */
  final def addTripleStatement(triple: TTriple): Iterable[RdfStreamRow] =
    handleHeader()
    val mainRow = RdfStreamRow(RdfStreamRow.Row.Triple(
      tripleToProto(triple)
    ))
    extraRowsBuffer.append(mainRow)

  /**
   * Add an RDF quad statement to the stream.
   * @param quad quad to add
   * @return iterable of stream rows
   */
  final def addQuadStatement(quad: TQuad): Iterable[RdfStreamRow] =
    handleHeader()
    val mainRow = RdfStreamRow(RdfStreamRow.Row.Quad(
      quadToProto(quad)
    ))
    extraRowsBuffer.append(mainRow)

  /**
   * Signal the start of a new (named) delimited graph in a GRAPHS stream.
   * @param graph graph node
   * @return iterable of stream rows
   */
  final def startGraph(graph: TNode): Iterable[RdfStreamRow] =
    if graph == null then
      startDefaultGraph()
    else
      handleHeader()
      val graphNode = graphNodeToProto(graph)
      val mainRow = RdfStreamRow(RdfStreamRow.Row.GraphStart(
        RdfGraphStart(graphNode)
      ))
      extraRowsBuffer.append(mainRow)

  /**
   * Signal the start of the default delimited graph in a GRAPHS stream.
   * @return iterable of stream rows
   */
  final def startDefaultGraph(): Iterable[RdfStreamRow] =
    handleHeader()
    val graphNode = RdfGraph(RdfGraph.Graph.DefaultGraph(RdfDefaultGraph()))
    val mainRow = RdfStreamRow(RdfStreamRow.Row.GraphStart(
      RdfGraphStart(graphNode)
    ))
    extraRowsBuffer.append(mainRow)

  /**
   * Signal the end of a delimited graph in a GRAPHS stream.
   * @return iterable of stream rows (always of length 1)
   */
  final def endGraph(): Iterable[RdfStreamRow] =
    if !emittedCompressionOptions then
      throw new RdfProtoSerializationError("Cannot end a delimited graph before starting one")
    Seq(RdfStreamRow(RdfStreamRow.Row.GraphEnd(RdfGraphEnd())))

  // *** 2. METHODS TO IMPLEMENT ***
  // *******************************
  // Triple statement deconstruction
  protected def getTstS(triple: TTriple): TNode
  protected def getTstP(triple: TTriple): TNode
  protected def getTstO(triple: TTriple): TNode

  // Quad statement deconstruction
  protected def getQstS(quad: TQuad): TNode
  protected def getQstP(quad: TQuad): TNode
  protected def getQstO(quad: TQuad): TNode
  protected def getQstG(quad: TQuad): TNode

  // Quoted triple term deconstruction
  protected def getQuotedS(triple: TQuoted): TNode
  protected def getQuotedP(triple: TQuoted): TNode
  protected def getQuotedO(triple: TQuoted): TNode

  /**
   * Turn an RDF node into its protobuf representation.
   *
   * Use the protected final inline make* methods in this class to create the nodes.
   *
   * @param node RDF node
   * @return option of RdfTerm
   * @throws RdfProtoSerializationError if node cannot be encoded
   */
  protected def nodeToProto(node: TNode): RdfTerm


  // *** 3. THE PROTECTED INTERFACE ***
  // **********************************
  protected final inline def makeIriNode(iri: String): RdfTerm =
    val iriEnc = iriEncoder.encodeIri(iri)
    RdfTerm(RdfTerm.Term.Iri(iriEnc))

  protected final inline def makeBlankNode(label: String): RdfTerm =
    RdfTerm(RdfTerm.Term.Bnode(label))

  protected final inline def makeSimpleLiteral(lex: String): RdfTerm =
    RdfTerm(RdfTerm.Term.Literal(
      RdfLiteral(lex, simpleLiteral)
    ))

  protected final inline def makeLangLiteral(lex: String, lang: String): RdfTerm =
    RdfTerm(RdfTerm.Term.Literal(
      RdfLiteral(lex, RdfLiteral.LiteralKind.Langtag(lang))
    ))

  protected final inline def makeDtLiteral(lex: String, dt: String): RdfTerm =
    RdfTerm(RdfTerm.Term.Literal(
      RdfLiteral(lex, iriEncoder.encodeDatatype(dt))
    ))

  protected final inline def makeTripleNode(triple: TQuoted): RdfTerm =
    RdfTerm(RdfTerm.Term.TripleTerm(quotedToProto(triple)))


  // *** 4. PRIVATE FIELDS AND METHODS ***
  // *************************************
  private val extraRowsBuffer = new ListBuffer[RdfStreamRow]()
  private val iriEncoder = new NameEncoder(options, extraRowsBuffer)
  private var emittedCompressionOptions = false

  private val lastSubject: LastNodeHolder[TNode] = new LastNodeHolder()
  private val lastPredicate: LastNodeHolder[TNode] = new LastNodeHolder()
  private val lastObject: LastNodeHolder[TNode] = new LastNodeHolder()
  private val lastGraph: LastNodeHolder[TNode] = new LastNodeHolder()

  private val nodeTermRepeat = RdfTerm(RdfTerm.Term.Repeat(RdfRepeat()))
  private val graphTermRepeat = RdfGraph(RdfGraph.Graph.Repeat(RdfRepeat()))
  private val simpleLiteral = RdfLiteral.LiteralKind.Simple(true)

  private def nodeToProtoWrapped(node: TNode, lastNodeHolder: LastNodeHolder[TNode]): RdfTerm =
    if options.useRepeat then
      lastNodeHolder.node match
        case oldNode if node == oldNode => nodeTermRepeat
        case _ =>
          lastNodeHolder.node = node
          nodeToProto(node)
    else
      nodeToProto(node)

  private def graphNodeToProto(node: TNode): RdfGraph =
    nodeToProto(node).term match
      case RdfTerm.Term.Iri(iri) => RdfGraph(RdfGraph.Graph.Iri(iri))
      case RdfTerm.Term.Literal(literal) => RdfGraph(RdfGraph.Graph.Literal(literal))
      case RdfTerm.Term.Bnode(bNode) => RdfGraph(RdfGraph.Graph.Bnode(bNode))
      case RdfTerm.Term.Empty => RdfGraph(RdfGraph.Graph.DefaultGraph(RdfDefaultGraph()))
      case _ => throw new RdfProtoSerializationError("Cannot encode node as a graph term")

  private def graphNodeToProtoWrapped(node: TNode): RdfGraph =
    if options.useRepeat then
      lastGraph.node match
        case oldNode if node == oldNode => graphTermRepeat
        case _ =>
          lastGraph.node = node
          graphNodeToProto(node)
    else
      graphNodeToProto(node)

  private def tripleToProto(triple: TTriple): RdfTriple =
    RdfTriple(
      s = nodeToProtoWrapped(getTstS(triple), lastSubject),
      p = nodeToProtoWrapped(getTstP(triple), lastPredicate),
      o = nodeToProtoWrapped(getTstO(triple), lastObject),
    )

  private def quadToProto(quad: TQuad): RdfQuad =
    RdfQuad(
      s = nodeToProtoWrapped(getQstS(quad), lastSubject),
      p = nodeToProtoWrapped(getQstP(quad), lastPredicate),
      o = nodeToProtoWrapped(getQstO(quad), lastObject),
      g = graphNodeToProtoWrapped(getQstG(quad)),
    )

  private def quotedToProto(quoted: TQuoted): RdfTriple =
    // ! No RdfRepeat support for inside of quoted triples.
    RdfTriple(
      s = nodeToProto(getQuotedS(quoted)),
      p = nodeToProto(getQuotedP(quoted)),
      o = nodeToProto(getQuotedO(quoted)),
    )

  private def handleHeader(): Unit =
    extraRowsBuffer.clear()
    if !emittedCompressionOptions then
      emittedCompressionOptions = true
      extraRowsBuffer.append(
        RdfStreamRow(RdfStreamRow.Row.Options(options))
      )

File: core-java/src/main/java/eu/ostrzyciel/jelly/core/ProtoEncoder.java
package eu.ostrzyciel.jelly.core;

import eu.ostrzyciel.jelly.core.internal.ProtoEncoderBase;
import eu.ostrzyciel.jelly.core.proto.v1.RdfStreamOptions;
import eu.ostrzyciel.jelly.core.proto.v1.RdfStreamRow;
import java.util.Collection;

/**
 * Base interface for RDF stream encoders.
 * @param <TNode> type of RDF nodes in the library
 */
public abstract class ProtoEncoder<TNode>
    extends ProtoEncoderBase<TNode>
    implements RowBufferAppender, ProtoHandler.AnyProtoHandler<TNode> {

    /**
     * Parameters passed to the Jelly encoder.
     * <p>
     * New fields may be added in the future, but always with a default value and in a sequential order.
     * However, it is still recommended to use named arguments when creating this object.
     *
     * @param options options for this stream (required)
     * @param enableNamespaceDeclarations whether to allow namespace declarations in the stream.
     *      If true, this will raise the stream version to 2 (Jelly 1.1.0). Otherwise,
     *      the stream version will be 1 (Jelly 1.0.0).
     * @param appendableRowBuffer buffer for storing stream rows that should go into a stream frame.
     *      The encoder will append the rows to this buffer.
     */
    public record Params(
        RdfStreamOptions options,
        boolean enableNamespaceDeclarations,
        Collection<RdfStreamRow> appendableRowBuffer
    ) {}

    /**
     * Whether namespace declarations are enabled for this encoder.
     */
    protected final boolean enableNamespaceDeclarations;

    /**
     * Buffer for storing stream rows that should go into a stream frame.
     */
    protected final Collection<RdfStreamRow> appendableRowBuffer;

    protected ProtoEncoder(ProtoEncoderConverter<TNode> converter, Params params) {
        super(params.options, converter);
        this.enableNamespaceDeclarations = params.enableNamespaceDeclarations;
        this.appendableRowBuffer = params.appendableRowBuffer;
    }

    /**
     * Add an RDF triple statement to the stream.
     * <p>
     * If your library does not support quad objects, use `addTripleStatement(s, p, o)` instead.
     *
     * @param triple triple to add
     * @throws RdfProtoSerializationError if the library does not support triple objects or
     *                                    if a serialization error occurs.
     */
    public final void addTripleStatement(TNode triple) {
        addTripleStatement(converter.getTstS(triple), converter.getTstP(triple), converter.getTstO(triple));
    }

    /**
     * Add an RDF triple statement to the stream.
     * @param subject subject
     * @param predicate predicate
     * @param object object
     * @since 2.9.0
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void addTripleStatement(TNode subject, TNode predicate, TNode object);

    /**
     * Add an RDF quad statement to the stream.
     * <p>
     * If your library does not support quad objects, use `addQuadStatement(s, p, o, g)` instead.
     *
     * @param quad quad to add
     * @throws RdfProtoSerializationError if the library does not support quad objects or
     *                                    if a serialization error occurs.
     */
    public final void addQuadStatement(TNode quad) {
        addQuadStatement(
            converter.getQstS(quad),
            converter.getQstP(quad),
            converter.getQstO(quad),
            converter.getQstG(quad)
        );
    }

    /**
     * Add an RDF quad statement to the stream.
     *
     * @param subject subject
     * @param predicate predicate
     * @param object object
     * @param graph graph
     * @since 2.9.0
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void addQuadStatement(TNode subject, TNode predicate, TNode object, TNode graph);

    /**
     * Signal the start of a new (named) delimited graph in a GRAPHS stream.
     * Null value is interpreted as the default graph.
     *
     * @param graph graph node
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void startGraph(TNode graph);

    /**
     * Signal the start of the default delimited graph in a GRAPHS stream.
     *
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void startDefaultGraph();

    /**
     * Signal the end of a delimited graph in a GRAPHS stream.
     *
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void endGraph();

    /**
     * Declare a namespace in the stream.
     * This is equivalent to the PREFIX directive in Turtle.
     *
     * @param name     short name of the namespace (without the colon)
     * @param iriValue IRI of the namespace
     * @throws RdfProtoSerializationError if a serialization error occurs
     */
    public abstract void declareNamespace(String name, String iriValue);
}
